packages/web-forms/src/components/OdkWebForm.vue

```diff
@@ -22,19 +25,54 @@ initializeForm(props.formXml, {
   }).catch(() => {}); // eslint-disable-line -- noop
 
 const handleSubmit = () => {
-	emit('submit');
+	if(odkForm.value?.validationState.violations?.length === 0){
+		emit('submit');
+	}
+	else{
+		submitPressed.value = true;
+		scrollToFirstInvalidQuestion();
+	}
+}
+
+provide('submitPressed', submitPressed);
+
+const formErrorMessage = computed(() => {
+	const violationLength = odkForm.value!.validationState.violations.length;
+
+	if(violationLength === 0) return '';
+	else if(violationLength === 1) return '1 question with error';
+	else return `${violationLength} questions with errors`;
+});
+
+const isInputElement = (e: HTMLElement): e is HTMLInputElement => e.tagName === 'INPUT';
+
+const scrollToFirstInvalidQuestion = () => {
+	document.getElementById(odkForm.value!.validationState.violations[0].nodeId + '_container')?.scrollIntoView({
+		behavior: 'smooth'
+	});
+	
+	// If first invalid element is a textbox then focus it.
+	const firstInvalidElement = document.getElementById(odkForm.value!.validationState.violations[0].nodeId);
+	if(firstInvalidElement && isInputElement(firstInvalidElement) && firstInvalidElement.type === 'text'){
+		firstInvalidElement.focus({preventScroll: true});
+	}
```

Comment:
For previous discussion context: [this thread](https://github.com/getodk/web-forms/pull/164#discussion_r1693632522) and [this other one](https://github.com/getodk/web-forms/pull/164#discussion_r1697520738).

> Still not sure what you are suggesting with "is a DOM scrolling API the way we want to handle programmatic in-form navigation?"

I wasn't suggesting anything in particular, it was an intentionally open-ended question. I was hoping to prompt discussion without biasing or coloring the discussion with any alternative details I might have in min. I was hoping that you would give some thought (if you hadn't already) to alternatives that you might find appealing.

But since you've asked for suggestions, here's a rough sketch of an approach I think might be better, both now (for usability and maintainability) and into known future work (where we know that not all questions will always be visible, and so won't necessarily be scroll targets in the first place):

1. Some form of shared/app global state indicating the question being targeted in a way that's generic enough it can be adapted to arbitrary question types. For instance, it could be as simple as setting the state to the targeted question's `nodeId`.

2. At the individual question level, some mechanism to react to that state change.

3. At the targeted question, some consistent way to instruct whatever component renders _its control_ that it should be activated (where "activated" brings it into view, performs whatever pertinent focus behavior, whatever other accessibility affordances we want to consider). This is something that will vary by control type, so it makes sense for _the control itself_ to handle it.

4. _Probably_ some way of making this state ephemeral, so that subsequent user interactions take precedence after the activation occurs.

5. Some way to signal to non-root ancestors when a descendant question is targeted. I'm not sure whether it makes sense for this communication to descend or ascend the hierarchy, but it will be important for paginated views. This is a good candidate to defer to later work, but it's something I want to consciously anticipate in the approach because we already know it will be coming.

Maybe we can do a pairing session to explore this? It seems likely there are some Vue APIs and common idioms that we could reach for. In other component libraries (React, Solid), I'd probably be thinking about utilizing their Context API or equivalent for coordination of global/question-local state, and then I'd probably be thinking about a base interface or "higher order component" for communication between the question and its control.

- - -

Since point 3 mentions that the activation behavior will likely be control-specific, I think it's worth mentioning that the special casing of text inputs here (a) hints at the same concern (glad we're both thinking about it!) but (b) confuses me as implemented (why wouldn't we want to focus other controls?). I'm also anticipating this will balloon into a lot of special cases that become harder to reason about and keep track of.

In general, I'm really nervous about the amount of implicit coupling this introduces now, and the ways that will be harder to manage over time.

As I said earlier, I did intentionally try to avoid making specific suggestions of alternatives. So I hope the rough sketch above won't be taken as too proscriptive about what alternatives might look like. And in that spirit, I'm happy to hear any other approaches you might have in mind as well.

- - -

More broadly, and maybe we can discuss this as a team tomorrow: it's worth thinking about whether we'd benefit from putting more design/discussion of approach upfront before diving into implementation. I think that's worked very well for the engine work we've landed over the last few months, and I think it would benefit the UI side as well. I also know it doesn't feel great to make progress on a big chunk of work, only to have significant questions about its approach come up in hindsight. Having more of a shared understanding of the expected design/approach would certainly help to reduce if not avoid that entirely.

Reply:
Removed "Fix errors" button and created https://github.com/getodk/web-forms/issues/189 to address that.